enzyme/test/Integration/ReverseMode/stl_list.cpp

```diff
@@ -0,0 +1,103 @@
+// FIXME: -O0 fails reverse mode (wrong result) https://github.com/EnzymeAD/Enzyme/pull/2370#issuecomment-3046307237
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O1 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O2 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O3 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+
+#include "../test_utils.h"
+
+#include <iostream>
+#include <list>
+
+
+struct S {
+    S(double r) : x(r) {};
+    double x = 0.0;
+};
+
+extern double __enzyme_fwddiff(void*, int, std::list<double>&, int, ...);
+extern double __enzyme_autodiff(void*, int, std::list<double>&, int, ...);
+extern double __enzyme_fwddiff(void*, int, std::list<S>&, int, ...);
+extern double __enzyme_autodiff(void*, int, std::list<S>&, int, ...);
+
+
+double test_iterate_list(std::list<double>& vals, double const & x) {
+    // iterate over list
+    double result = 0.0;
+    for (const auto& val : vals) {
```

Comment:
@wsmoses I think reverse mode with list iteration does not yet work. My tests below fail for it, returning random/undefined values.

Reply:
fixed, need to read `dx` not the returned value in reverse mode.